modules/ttdBidAdapter.js

```diff
@@ -0,0 +1,602 @@
+          return false;
+        }
+      }
+    }
+
+    const keywords = utils.deepAccess(bid, 'params.keywords');
+    if (keywords) {
+      if (!Array.isArray(keywords)) {
+        utils.logWarn(BIDDER_CODE + ': params.keywords should be an array of string values');
+        return false;
+      }
+      for (let i = 0; i < keywords.length; i++) {
+        if (typeof keywords[i] !== 'string') {
+          utils.logWarn(BIDDER_CODE + ': params.keywords should be an array of string values');
+          return false;
+        }
+        if (keywords[i].length > 128) {
+          utils.logWarn(BIDDER_CODE + ': params.keywords must only contain strings that are 128 characters or less');
+          return false;
+        }
+      }
+      if (keywords.length > 10) {
+        utils.logWarn(BIDDER_CODE + ': params.keywords must can only contain a max of 10 keywords');
+        return false;
+      }
+    }
+
+    return true;
+  },
+
+  /**
+   * Make a server request from the list of BidRequests.
+   *
+   * @param {BidRequest[]} an array of validBidRequests
+   * @param {*} bidderRequest
+   * @return {ServerRequest} Info describing the request to the server.
+   */
+  buildRequests: function (validBidRequests, bidderRequest) {
+    utils.logInfo('validBidRequests:', JSON.stringify(validBidRequests))
+    utils.logInfo('bidderRequest:', JSON.stringify(bidderRequest))
+
+    let topLevel = {
+      id: bidderRequest.auctionId,
+      imp: validBidRequests.map(bid => getImpression(bid)),
+      site: getSite(bidderRequest),
+      device: getDevice(),
+      user: getUser(bidderRequest),
+      at: 1,
+      cur: ['USD'],
+      regs: getRegs(bidderRequest),
+      source: getSource(validBidRequests),
+      ext: getExt(bidderRequest)
+    }
+
+    let url = BIDDER_ENDPOINT + bidderRequest.bids[0].params.supplySourceId;
+
+    let serverRequest = {
+      method: 'POST',
+      url: url,
+      data: topLevel,
+      options: {
+        withCredentials: true
+      }
+    };
+    utils.logInfo('buildRequests serverRequest:', JSON.stringify(serverRequest))
+    return serverRequest;
+  },
+
+  /**
+   * Format responses as Prebid bid responses
+   *
+   * Each bid can have the following elements:
+   * - requestId (required)
+   * - cpm (required)
+   * - width (required)
+   * - height (required)
+   * - ad (required)
+   * - ttl (required)
+   * - creativeId (required)
+   * - netRevenue (required)
+   * - currency (required)
+   * - vastUrl
+   * - vastImpUrl
+   * - vastXml
+   * - dealId
+   *
+   * @param {ttdResponseObj} bidResponse A successful response from ttd.
+   * @param {ServerRequest} serverRequest The result of buildRequests() that lead to this response.
+   * @return {Bid[]} An array of formatted bids.
+  */
+  interpretResponse: function (response, serverRequest) {
+    utils.logInfo('interpretResponse-bidResponse:', JSON.stringify(response))
+    utils.logInfo('interpretResponse-serverRequest:', JSON.stringify(serverRequest))
```

Comment:
Could all logInfo's used in this PR be reduced to only crucial communication to the user?

Reply:
removed all debug logging